simulated_beacon/simulated_beacon.go

```diff
@@ -0,0 +1,223 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package simulated_beacon
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/beacon/engine"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/eth"
+	"github.com/ethereum/go-ethereum/eth/catalyst"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/node"
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+// withdrawals implements a FIFO queue which holds withdrawals that are pending inclusion
+type withdrawals struct {
+	pending []*types.Withdrawal
+	next    uint64
+}
+
+// pop removes up to 10 withdrawals from the queue
+func (w *withdrawals) pop() []*types.Withdrawal {
+	var popCount int
+	if len(w.pending) >= 10 {
+		popCount = 10
+	} else {
+		popCount = len(w.pending)
+	}
+
+	popped := make([]*types.Withdrawal, popCount)
+	copy(popped[:], w.pending[0:popCount])
+	w.pending = append([]*types.Withdrawal{}, w.pending[popCount:]...)
+	return popped
+}
+
+// add adds a withdrawal to the queue
+func (w *withdrawals) add(withdrawal *types.Withdrawal) error {
+	if withdrawal.Index < w.next {
+		return fmt.Errorf("withdrawal has index (%d) less than or equal to latest received withdrawal index (%d)", withdrawal.Index, w.next-1)
+	}
+	w.next = withdrawal.Index + 1
+	w.pending = append(w.pending, withdrawal)
+	return nil
+}
+
+type SimulatedBeacon struct {
+	ctx          context.Context
+	cancel       context.CancelFunc
+	eth          *eth.Ethereum
+	period       time.Duration
+	withdrawals  withdrawals
+	feeRecipient common.Address
+	// mu controls access to the feeRecipient/withdrawals which can be modified by the dev-mode RPC API methods
+	mu sync.Mutex
+}
+
+func NewSimulatedBeacon(eth *eth.Ethereum) *SimulatedBeacon {
+	chainConfig := eth.APIBackend.ChainConfig()
+	if chainConfig.Dev == nil {
+		log.Crit("incompatible pre-existing chain configuration")
+	}
+
+	return &SimulatedBeacon{
+		eth:          eth,
+		period:       time.Duration(chainConfig.Dev.Period) * time.Second,
+		withdrawals:  withdrawals{[]*types.Withdrawal{}, 0},
+		feeRecipient: common.Address{},
+	}
+}
+
+// Start invokes the SimulatedBeacon life-cycle function in a goroutine
+func (c *SimulatedBeacon) Start() error {
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+	go c.loop()
+	return nil
+}
+
+// Stop halts the SimulatedBeacon service
+func (c *SimulatedBeacon) Stop() error {
+	c.cancel()
+	return nil
+}
```

Comment:
We usually do it like this 
```golang
func (c *SimulatedBeacon) Start() error {
	c.shutdownCh = make(chan struct{})
	go c.loop()
	return nil
}

// Stop halts the SimulatedBeacon service
func (c *SimulatedBeacon) Stop() error {
	close(c.shutdownCh)
	return nil
}
```
Then we only need to keep track of one thing -- not both a `ctx` and  `CancelFunc`

Reply:
Okay :+1: .  I think my use of a context here is a carry-over from when I was invoking RPC methods through the engine API external handlers (which require a context as a parameter).